// ng-typing/esm2015/ng-typing.js
import { Component, Input, Output, EventEmitter, NgModule } from '@angular/core';

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
class TypingComponent {
    constructor() {
        this.referenceSpeed = 60;
        this.speedDeviation = 30;
        this.typo = true;
        this.typoProbability = 5;
        this.onCompleted = new EventEmitter();
        this.chrs = [];
        this.position = 0;
        this.displayedChrs = '';
        this.progress = false;
    }
    /**
     * @return {?}
     */
    ngOnInit() {
        this.show();
    }
    /**
     * @param {?=} message
     * @return {?}
     */
    show(message) {
        if (message) {
            this.message = message;
        }
        if (!this.progress && this.message && this.message.trim().length > 0) {
            this.progress = true;
            this.initParameters();
            if (this.referenceSpeed == 0) {
                this.displayedChrs = this.message;
            }
            else {
                this.chrs = this.message.trim().split('');
                setTimeout(this.displayCharacter.bind(this), this.randomSpeed());
            }
        }
    }
    /**
     * @return {?}
     */
    initParameters() {
        this.chrs = [];
        this.position = 0;
        this.displayedChrs = '';
    }
    /**
     * @return {?}
     */
    displayCharacter() {
        if (this.doTypo()) {
            this.displayedChrs += this.getRandomCharacter();
            setTimeout(this.correctTypo.bind(this), TypingComponent.TYPO_FIX_SPEED);
        }
        else {
            this.displayedChrs += this.chrs[this.position++];
            if (this.position == this.chrs.length) {
                this.onCompleted.emit();
                this.progress = false;
            }
            else {
                setTimeout(this.displayCharacter.bind(this), this.randomSpeed());
            }
        }
    }
    /**
     * @return {?}
     */
    getRandomCharacter() {
        return TypingComponent.RANDOM_CHARACTERS
            .charAt(Math.floor(Math.random() * TypingComponent.RANDOM_CHARACTERS.length));
    }
    /**
     * @return {?}
     */
    correctTypo() {
        this.displayedChrs = this.displayedChrs.slice(0, -1);
        setTimeout(this.displayCharacter.bind(this), this.randomSpeed());
    }
    /**
     * @return {?}
     */
    randomSpeed() {
        const /** @type {?} */ deviation = Math.floor(Math.random() * Math.floor(this.speedDeviation));
        const /** @type {?} */ direction = Math.round(Math.random()) == 0 ? 1 : -1;
        return this.referenceSpeed + (deviation * direction);
    }
    /**
     * @return {?}
     */
    doTypo() {
        if (this.position > 2 && !!this.typo) {
            this.typoProbability = Math.max(Math.min(TypingComponent.MAX_TYPO_PROBABILITY, Math.round(this.typoProbability)), 0);
            return Math.floor(Math.random() * 100) < this.typoProbability;
        }
        return false;
    }
}
TypingComponent.MAX_TYPO_PROBABILITY = 80;
TypingComponent.TYPO_FIX_SPEED = 200;
TypingComponent.RANDOM_CHARACTERS = 'ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz';
TypingComponent.decorators = [
    { type: Component, args: [{
                selector: 'typing',
                template: `{{displayedChrs}}`
            },] },
];
/** @nocollapse */
TypingComponent.ctorParameters = () => [];
TypingComponent.propDecorators = {
    "message": [{ type: Input },],
    "referenceSpeed": [{ type: Input },],
    "speedDeviation": [{ type: Input },],
    "typo": [{ type: Input },],
    "typoProbability": [{ type: Input },],
    "onCompleted": [{ type: Output, args: ['onCompleted',] },],
};

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
class TypingModule {
}
TypingModule.decorators = [
    { type: NgModule, args: [{
                declarations: [TypingComponent],
                exports: [TypingComponent]
            },] },
];
/** @nocollapse */
TypingModule.ctorParameters = () => [];

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
/**
 * Generated bundle index. Do not edit.
 */

export { TypingModule, TypingComponent };
